components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt

```diff
@@ -227,13 +229,15 @@ class Session(
      */
     var canGoBack: Boolean by Delegates.observable(false) { _, old, new ->
         notifyObservers(old, new) { onNavigationStateChanged(this@Session, new, canGoForward) }
+        store?.syncDispatch(UpdateNavigationStateAction(id, NavigationState(canGoBack, canGoForward)))
     }
 
     /**
      * Navigation state, true if there's an history item to go forward to, otherwise false.
      */
     var canGoForward: Boolean by Delegates.observable(false) { _, old, new ->
         notifyObservers(old, new) { onNavigationStateChanged(this@Session, canGoBack, new) }
+        store?.syncDispatch(UpdateNavigationStateAction(id, NavigationState(canGoBack, canGoForward)))
```

Comment:
I wonder if always dispatching a combined `NavigationState` is the best solution when in the future we will remove it from `Session` and only have it on `SessionState`. I did the same but then looked at `EngineObserver` which gets a nullable `canGoBack` and `canGoForward` (one level lower Gecko also only invokes `onCanGoForward` or `onCanGoBack`).
Once we get rid of `Session` we will not have the combined state and only the `back` or `forward` change to dispatch. In my local version I then changed the action to have nullable fields and only dispatch what actually changed and only update that field in the reducer. But at this point we could also just have two actions. What do you think?

Reply:
Heh, I actually started with that and then changed it to using a single `NavigationState`. This seemed cleaner to me, as it's a bit easier to read than adding two more boolean states to the `ContentState` directly.

> Once we get rid of Session we will not have the combined state and only the back or forward change to dispatch. 

Yeah, that's a good point. We're using independent / separate actions everywhere. So, I will change this to separate actions as well.

@pocmo btw, I noticed that restored sessions lose their history i.e. navigation doesn't work once a session was restored. I can't seem to remember if this worked before? So, open a tab, follow a few links, save, close and open the app. We should be able to go/navigate back but can't. Do we have an issue on file (GV or A-C)? Kinda confused how we missed this, probably just me :).

Reply:
> @pocmo btw, I noticed that restored sessions lose their history i.e. navigation doesn't work once a session was restored. I can't seem to remember if this worked before? 

I am pretty sure this worked and that I have been doing this a bunch of times to test something not that long ago. If I remember correctly then the only way this could have worked is when GV invokes the callbacks on restore, I think? In theory this sounds like a simple test we could add to the on-device tests of the `browser-engine-gecko-nightly` component - and then maybe use that to bisect if that worked earlier?

Reply:
@pocmo OK, I filed an issue: https://github.com/mozilla-mobile/android-components/issues/6783 - It also doesn't work in Fenix release. So, the regression isn't too recent.